hyper-capi/src/http_types.rs

```diff
@@ -0,0 +1,132 @@
+use hyper::{Body, HeaderMap, Method, Request, Response, Uri};
+use hyper::header::{HeaderName, HeaderValue};
+use libc::size_t;
+use std::ffi::c_void;
+
+use crate::task::{AsTaskType, TaskType};
+use crate::{hyper_code, IterStep};
+
+// ===== impl Request =====
+
+ffi_fn! {
+    fn hyper_request_new() -> *mut Request<Body> {
+        Box::into_raw(Box::new(Request::new(Body::empty())))
+    }
+}
+
+ffi_fn! {
+    fn hyper_request_free(req: *mut Request<Body>) {
+        drop(unsafe { Box::from_raw(req) });
+    }
+}
+
+ffi_fn! {
+    fn hyper_request_set_method(req: *mut Request<Body>, method: *const u8, method_len: size_t) -> hyper_code {
+        let bytes = unsafe {
+            std::slice::from_raw_parts(method, method_len as usize)
+        };
+        match Method::from_bytes(bytes) {
+            Ok(m) => {
+                *unsafe { &mut *req }.method_mut() = m;
+                hyper_code::Ok
+            },
+            Err(_) => {
+                hyper_code::Kaboom
+            }
+        }
+    }
+}
+
+ffi_fn! {
+    fn hyper_request_set_uri(req: *mut Request<Body>, uri: *const u8, uri_len: size_t) -> hyper_code {
+        let bytes = unsafe {
+            std::slice::from_raw_parts(uri, uri_len as usize)
+        };
+        match Uri::from_maybe_shared(bytes) {
+            Ok(u) => {
+                *unsafe { &mut *req }.uri_mut() = u;
+                hyper_code::Ok
+            },
+            Err(_) => {
+                hyper_code::Kaboom
+            }
+        }
+    }
+}
+
+ffi_fn! {
+    fn hyper_request_headers(req: *mut Request<Body>) -> *mut HeaderMap {
+        unsafe { &mut *req }.headers_mut()
```

Comment:
Is this sound from a rust memory model perspective? We're returning a pointer to an inner field not accessed through an `UnsafeCell`. I'm not sure how the full aliasing model of rust works with pointers; is it an issue that the mutable reference we got that pointer from has gone out of scope here? Or is it fine as long as we don't dereference `req` again until we're done with this `*mut HeaderMap`? Do we run into theoretical soundness issues if C code asks for the headers first and then asks for e.g. `hyper_request_set_uri`?

This is all a bit unlikely unless LTO crosses between C and Rust, of course, but with LLVM that's a possibility.

You could of course sidestep these soundness issues by having `hyper_headers_set` take `*mut Request<Body>` instead.

Reply:
I've checked with @pnkfelix, and he believes that what this is doing is currently safe. It *would* be invalid for a user in C to call this multiple times and access the `*mut HeaderMap` simultaneously, but simply returning this pointer should be sound (and shouldn't require an `UnsafeCell`).